sdk/nodejs/output.ts

```diff
@@ -460,6 +472,66 @@ function getResourcesAndDetails<T>(allOutputs: Output<Unwrap<T>>[]): [Resource[]
     return [allResources, isKnown, isSecret];
 }
 
+/**
+ * Unknown represents a value that is unknown.
+ */
+export class Unknown {
+     /**
+     * @internal
+     * A private field to help with RTTI that works in SxS scenarios.
+     *
+     * This is internal instead of being truly private, to support mixins and our serialization model.
+     */
+    // tslint:disable-next-line:variable-name
+    public readonly __pulumiUnknown: boolean = true;
+
+    /**
+     * Returns true if the given object is an instance of Unknown. This is designed to work even when
+     * multiple copies of the Pulumi SDK have been loaded into the same process.
+     */
+    public static isInstance(obj: any): obj is Unknown {
+        return utils.isInstance(obj, "__pulumiUnknown");
+    }
+}
+
+/**
+ * unknown is the singleton unknown value.
+ */
+export const unknown = new Unknown();
+
+/**
+ * isUnknown returns true if the given value is unknown.
+ */
+export function isUnknown(val: any): boolean {
+    return Unknown.isInstance(val);
+}
+
+/**
+ * containsUnknowns returns true if the given value is or contains unknown values.
+ */
+export function containsUnknowns(value: any): boolean {
+    function impl(val: any, seen: Set<any>): boolean {
+        if (val === null || typeof val !== "object") {
+            return false;
+        }
+        else if (isUnknown(val)) {
+            return true;
+        }
+        else if (seen.has(val)) {
+            return false;
+        }
+        else if (val instanceof Array) {
+            seen.add(val);
+            return val.some(e => impl(e, seen));
+        }
+        else {
+            seen.add(val);
+            return Object.keys(val).some(k => impl(val[k], seen));
+        }
+    }
+    return impl(value, new Set<any>());
```

Comment:
in general, our preference is to define local functions after actual function code.

Reply:
...gross, but ok.